src/poetry/console/commands/run.py

```diff
@@ -44,7 +45,25 @@ def _module(self) -> Module:
 
         return module
 
-    def run_script(self, script: str | dict[str, str], args: str) -> int:
+    def run_script(self, script: str | dict[str, str], args: list[str]) -> int:
+        """Runs an entry point script defined in the section ``[tool.poetry.scripts]``.
+
+        When a script exists in the venv bin folder, i.e. after ``poetry install``,
+        then ``sys.argv[0]`` must be set to the full path of the executable, so
+        ``poetry run foo`` and ``poetry shell``, ``foo`` have the same ``sys.argv[0]``
+        that points to the full path.
+
+        Otherwise (when an entry point script does not exist), ``sys.argv[0]`` is the
+        script name only, i.e. ``poetry run foo`` has ``sys.argv == ['foo']``.
+        """
+        for script_dir in self.env.script_dirs:
+            script_path = script_dir / args[0]
+            if WINDOWS:
+                script_path = script_path.with_suffix(".cmd")
+            if script_path.exists():
+                args[0] = str(script_path)
+                break
```

Comment:
Never thought about that. It's safe as long nobody else relies on args being unchanged. I think that's the case, here. However, looking into the cleo code it seems it just returns it's internal list (not a copy) and changing it doesn't seem correct.

Please go ahead with the copy approach. That way we stay safe even if cleo decides to return a tuple instead of a list.

Reply:
Great, just done!